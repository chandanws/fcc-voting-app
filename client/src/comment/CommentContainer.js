import React, { Component } from "react";
import CommentComponent from "./CommentComponent";
import {
  toggleElementInArray,
  toggleObjectsInArray,
  changeValueInArrayOfObjects
} from "../helpers/helpers";
import { connect } from "react-redux";
import { fetchComments } from "../actions/commentActions";

class CommentContainer extends Component {
  // comments will later be moved to redux state
  // closed comments and open reply boxes will move to new state
  // open replies => [ {id: 1, value: "something"}, {id: 2, value: "something" }]
  state = {
    closed: [],
    openReplies: []
  };

  componentDidMount() {
    this.props.fetchComments(this.props.poll_id);
  }

  toggleOpenReplies = id => {
    const newArray = toggleObjectsInArray(this.state.openReplies, id);
    this.setState({
      openReplies: newArray
    });
    console.log("Test", newArray);
  };

  toggleTree = id => {
    const newArray = toggleElementInArray(this.state.closed, id);
    this.setState({
      closed: newArray
    });
  };

  handleReplyState = id => event => {
    const newArray = changeValueInArrayOfObjects(
      this.state.openReplies,
      id,
      event.target.value
    );
    this.setState({
      openReplies: newArray
    });
  };

  render() {
    console.log("Comments", this.props.comments);
    return (
      <div style={{ marginLeft: "-2rem" }}>
        {this.props.comments.data[0] && (
          <Comments
            openReplies={this.state.openReplies}
            toggleOpenReplies={this.toggleOpenReplies}
            handleReplyState={this.handleReplyState}
            closed={this.state.closed}
            toggleTree={this.toggleTree}
            comments={this.props.comments.data}
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  comments: state.comments
});

const mapDispatchToProps = dispatch => ({
  fetchComments: poll_id => dispatch(fetchComments(poll_id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CommentContainer);

export const Comments = props => {
  return (
    <div className="comments-branch">
      {props.comments.map((comment, index) => {
        return (
          <CommentComponent
            key={index}
            openReplies={props.openReplies}
            toggleOpenReplies={props.toggleOpenReplies}
            handleReplyState={props.handleReplyState}
            closed={props.closed}
            toggleTree={props.toggleTree}
            comment={comment}
          />
        );
      })}
    </div>
  );
};
