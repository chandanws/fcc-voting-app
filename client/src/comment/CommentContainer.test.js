import React from "react";
import CommentContainer from "./CommentContainer";
import { shallow } from "enzyme";

describe("<CommentContainer />", () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<CommentContainer />);
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("toggleOpenReplies(id)", () => {
    beforeEach(() => {
      wrapper.setState({ openReplies: [{ id: 1, value: "nothing" }] });
    });
    it("should change state.openReplies", () => {
      wrapper.instance().toggleOpenReplies(1);
      wrapper.update();
      expect(wrapper.state().openReplies).toEqual([]);
      wrapper.instance().toggleOpenReplies(2);
      wrapper.update();
      expect(wrapper.state().openReplies).toEqual([{ id: 2, value: "" }]);
      wrapper.instance().toggleOpenReplies(1);
      wrapper.update();
      expect(wrapper.state().openReplies).toEqual([
        { id: 2, value: "" },
        { id: 1, value: "" }
      ]);
    });
  });
});
