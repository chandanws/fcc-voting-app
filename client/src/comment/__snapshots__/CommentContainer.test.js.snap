// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentContainer /> should render correctly 1`] = `
<div
  style={
    Object {
      "marginLeft": "-2rem",
    }
  }
>
  <Component
    closed={Array []}
    comments={
      Array [
        Object {
          "body": "this is random body",
          "children": Array [],
          "id": 1,
          "username": "user1",
        },
        Object {
          "body": "this is random body",
          "children": Array [
            Object {
              "body": "this is random body",
              "children": Array [
                Object {
                  "body": "this is random body",
                  "children": Array [],
                  "id": 30,
                  "username": "user1",
                },
                Object {
                  "body": "this is random body",
                  "children": Array [],
                  "id": 7,
                  "username": "user1",
                },
              ],
              "id": 5,
              "username": "user3",
            },
            Object {
              "body": "this is random body",
              "children": Array [],
              "id": 4,
              "username": "user1",
            },
          ],
          "id": 2,
          "username": "user2",
        },
        Object {
          "body": "this is random body",
          "children": Array [
            Object {
              "body": "this is random body",
              "children": Array [],
              "id": 32,
              "username": "user1",
            },
          ],
          "id": 31,
          "username": "user1",
        },
        Object {
          "body": "this is random body",
          "children": Array [],
          "id": 33,
          "username": "user1",
        },
      ]
    }
    handleReplyState={[Function]}
    toggleOpenReplies={[Function]}
    toggleTree={[Function]}
  />
</div>
`;
